core/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSourceParser.java

```diff
@@ -0,0 +1,210 @@
+     * @param targetValueType
+     *            the target type of the final value output by the aggregation
+     * @param otherOptions
+     *            a {@link Map} containing the extra options parsed by the
+     *            {@link #token(String, String, org.elasticsearch.common.xcontent.XContentParser.Token,
+     *             XContentParser, ParseFieldMatcher, Map)}
+     *            method
+     * @return the created factory
+     */
+    protected abstract MultiValuesSourceAggregatorBuilder<VS, ?> createFactory(String aggregationName, ValuesSourceType valuesSourceType,
+            ValueType targetValueType, Map<ParseField, Object> otherOptions);
+
+    @Override
+    protected MultiValuesSourceAggregatorBuilder<VS, ?> createBuilder(String aggregationName, final ValueType parsedValueType,
+            final String parsedFormat, final Object parsedMissing,
+            final DateTimeZone parsedTimeZone, Map<ParseField, Object> otherOptions) {
+        MultiValuesSourceAggregatorBuilder<VS, ?> factory = createFactory(aggregationName, this.valuesSourceType, this.targetValueType,
+            otherOptions);
+        if (fields != null) {
+            factory.fields(fields);
+        }
+        if (scripts != null) {
+            factory.scripts(scripts);
+        }
+        if (parsedValueType != null) {
+            factory.valueType(parsedValueType);
+        }
+        if (parsedFormat != null) {
+            factory.format(parsedFormat);
+        }
+        if (parsedMissing != null) {
+            factory.missing(parsedMissing);
+        }
+        if (parsedTimeZone != null) {
+            factory.timeZone(parsedTimeZone);
+        }
+        return factory;
+    }
+
+    /**
+     * Parse field and scripts as arrays
+     */
+    @Override
+    protected boolean internalToken(String aggregationName, String currentFieldName, XContentParser.Token token, XContentParser parser,
+                                    ParseFieldMatcher parseFieldMatcher, Map<ParseField, Object> otherOptions) throws IOException {
+        if (scriptable && token == XContentParser.Token.START_ARRAY) {
+            if (parseFieldMatcher.match(currentFieldName, ScriptField.SCRIPT)) {
+                if (scripts == null) {
+                    scripts = new HashMap<>();
+                } else {
+                    scripts.clear();
+                }
+                while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                    // since each script must have a name, this must be an array of objects
+                    if (token == XContentParser.Token.START_OBJECT) {
+                        parser.nextToken();
+                        parseScriptAndAdd(aggregationName, currentFieldName, parser, parseFieldMatcher);
+                    } else {
+                        return false;
+                    }
+                }
+            } else if ("field".equals(currentFieldName)) {
```

Comment:
Can we use ParseField here? In fact, could we take this opportunity to update the SingleValuesSourceParser and ValuesSourceParser to use ParseField too?